src/darwin/Framework/CHIP/MTROperationalCredentialsDelegate.mm

```diff
@@ -112,9 +115,87 @@
+    }
+}
+
+CHIP_ERROR MTROperationalCredentialsDelegate::CallbackGenerateNOCChain(const chip::ByteSpan & csrElements,
+    const chip::ByteSpan & csrNonce, const chip::ByteSpan & csrElementsSignature, const chip::ByteSpan & attestationChallenge,
+    const chip::ByteSpan & DAC, const chip::ByteSpan & PAI,
+    chip::Callback::Callback<chip::Controller::OnNOCChainGeneration> * onCompletion)
+{
+    mOnNOCCompletionCallback = onCompletion;
+
+    TLVReader reader;
+    reader.Init(csrElements);
+
+    if (reader.GetType() == kTLVType_NotSpecified) {
+        ReturnErrorOnFailure(reader.Next());
+    }
+
+    VerifyOrReturnError(reader.GetType() == kTLVType_Structure, CHIP_ERROR_WRONG_TLV_TYPE);
+    VerifyOrReturnError(reader.GetTag() == AnonymousTag(), CHIP_ERROR_UNEXPECTED_TLV_ELEMENT);
+
+    TLVType containerType;
+    ReturnErrorOnFailure(reader.EnterContainer(containerType));
+    ReturnErrorOnFailure(reader.Next(kTLVType_ByteString, TLV::ContextTag(1)));
+
+    chip::ByteSpan csr(reader.GetReadPoint(), reader.GetLength());
+    reader.ExitContainer(containerType);
+
+    CSRInfo * csrInfo = [[CSRInfo alloc] initWithNonce:AsData(csrNonce)
+                                              elements:AsData(csrElements)
+                                     elementsSignature:AsData(csrElementsSignature)
+                                                   csr:AsData(csr)];
+
+    chip::ByteSpan certificationDeclarationSpan;
+    chip::ByteSpan attestationNonceSpan;
+    uint32_t timestampDeconstructed;
+    chip::ByteSpan firmwareInfoSpan;
+    chip::Credentials::DeviceAttestationVendorReservedDeconstructor vendorReserved;
+
+    chip::Controller::CommissioningParameters commissioningParameters = mAutoCommissioner->GetCommissioningParameters();
+    ReturnErrorOnFailure(chip::Credentials::DeconstructAttestationElements(commissioningParameters.GetAttestationElements().Value(),
+        certificationDeclarationSpan, attestationNonceSpan, timestampDeconstructed, firmwareInfoSpan, vendorReserved));
+
+    AttestationInfo * attestationInfo =
+        [[AttestationInfo alloc] initWithChallenge:AsData(attestationChallenge)
+                                             nonce:AsData(commissioningParameters.GetAttestationNonce().Value())
+                                          elements:AsData(commissioningParameters.GetAttestationElements().Value())
+                                 elementsSignature:AsData(commissioningParameters.GetAttestationSignature().Value())
+                                               dac:AsData(DAC)
+                                               pai:AsData(PAI)
+                          certificationDeclaration:AsData(certificationDeclarationSpan)
+                                      firmwareInfo:AsData(firmwareInfoSpan)];
+
+    [mNocChainIssuer onNOCChainGenerationNeeded:csrInfo attestationInfo:attestationInfo];
```

Comment:
This should be taking a callback block to call with the results instead of going through the controller and whatnot, I would think.  Need to be a little careful about the API shape here.

Reply:
Done.